File: services/shadowit-service/internal/tracing/tracing.go
// Package tracing wires this service into the shared Tempo backend
// (infra/tempo) via OpenTelemetry. Before this, the only cross-service
// visibility was Prometheus counters per service — no way to see, for one
// slow or failed request, the whole hop chain (admin-api -> dlp-service ->
// ClamAV, or admin-api -> ai-service -> the LLM provider) in one place.
//
// Every service in this stack that talks to another service should call
// Init at startup and wrap its outbound http.Client's Transport with
// otelhttp.NewTransport, so the trace context (the W3C traceparent header)
// actually propagates end-to-end rather than each service starting its own
// disconnected trace.
package tracing

import (
	"context"
	"log"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

// Init configures the global TracerProvider and propagator for serviceName,
// exporting spans to OTEL_EXPORTER_OTLP_ENDPOINT (default: Tempo's OTLP/HTTP
// receiver on the shared Docker network). Returns a shutdown func to call
// on exit — flushes any buffered spans rather than dropping them.
//
// Never fails startup: an unreachable collector means spans queue and
// eventually drop, not that the service refuses to boot. Observability
// infrastructure being down must never take the product down with it.
func Init(serviceName string) func(context.Context) error {
	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		endpoint = "tempo:6318"
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(endpoint),
		otlptracehttp.WithInsecure(), // plain HTTP inside the Docker network, same trust boundary as every other internal call here
	)
	if err != nil {
		// Exporter construction itself doesn't dial anything — this only
		// fails on bad config (a malformed endpoint), which is a real setup
		// mistake worth a log line, but still not worth crashing the
		// service over. Leaving the global TracerProvider unset here means
		// every span created afterward is OTel's own no-op tracer — cheap,
		// safe, and exactly what "tracing is unavailable" should look like.
		log.Printf("tracing: could not init OTLP exporter (%v) — spans will not be exported", err)
		return func(context.Context) error { return nil }
	}

	res, _ := resource.New(ctx,
		resource.WithAttributes(semconv.ServiceName(serviceName)),
	)

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
		// Every request traced, not sampled — this stack's traffic volume
		// doesn't need head-based sampling yet, and under-sampling would
		// undermine the exact thing this exists for: seeing the one request
		// that actually failed.
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown
}

File: services/dlp-service/app/tracing.py
"""Wires this service into the shared Tempo backend (infra/tempo) via
OpenTelemetry — the Python counterpart to services/admin-api/internal/tracing.
Every service that talks to another service instruments both its inbound
(FastAPI) and outbound (httpx) traffic, so the trace context (the W3C
traceparent header) actually propagates end-to-end into one connected
trace instead of each service starting its own.
"""

from __future__ import annotations

import logging
import os

logger = logging.getLogger(__name__)


def setup_tracing(app, service_name: str) -> None:
    """Instruments `app` (a FastAPI instance) for both inbound request spans
    and outbound httpx spans. Never raises — an unreachable collector or a
    missing OTel package must not stop the service from starting; spans
    just don't get created/exported, which is the correct degraded state,
    not a crash.
    """
    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
        from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
        from opentelemetry.sdk.resources import SERVICE_NAME, Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        # OTEL_EXPORTER_OTLP_ENDPOINT is conventionally a bare scheme://host:port
        # (matches the Go services' OTLP endpoint convention) — /v1/traces is
        # appended here since passing `endpoint` directly to OTLPSpanExporter
        # skips the auto-path-append the env-var-only config path would do.
        base = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://tempo:6318")
        endpoint = base.rstrip("/") + "/v1/traces"

        provider = TracerProvider(resource=Resource.create({SERVICE_NAME: service_name}))
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint)))
        trace.set_tracer_provider(provider)

        FastAPIInstrumentor.instrument_app(app)
        HTTPXClientInstrumentor().instrument()

        logger.info("tracing: exporting spans for %s to %s", service_name, endpoint)
    except Exception as exc:  # noqa: BLE001 - deliberately broad, see docstring
        logger.warning("tracing: could not initialize OpenTelemetry (%s) — continuing without it", exc)
